roachtest: tpcc/w=100/nodes=3/chaos=true failed

roachtest.tpcc/w=100/nodes=3/chaos=true [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3431372&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3431372&tab=artifacts#/tpcc/w=100/nodes=3/chaos=true) on master @ [394e35cbbb7ae466383c3ee6f2bbcc76857afeed](https://github.com/cockroachdb/cockroach/commits/394e35cbbb7ae466383c3ee6f2bbcc76857afeed):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/tpcc/w=100/nodes=3/chaos=true/run_1
	monitor.go:128,tpcc.go:279,tpcc.go:689,test_runner.go:777: monitor failure: unexpected node event: 1: dead (exit status 134)
		(1) attached stack trace
		  -- stack trace:
		  | main.(*monitorImpl).WaitE
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/monitor.go:116
		  | main.(*monitorImpl).Wait
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/monitor.go:124
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runTPCC
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/tpcc.go:279
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.registerTPCC.func10
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/tpcc.go:689
		  | main.(*testRunner).runTest.func2
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:777
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1371
		Wraps: (2) monitor failure
		Wraps: (3) unexpected node event: 1: dead (exit status 134)
		Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *errors.errorString

	cluster.go:1249,context.go:89,cluster.go:1237,test_runner.go:866: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-3431372-1631513463-02-n4cpu4 --oneshot --ignore-empty-nodes: exit status 1 1: dead (exit status 134)
		2: 21473
		3: 21621
		4: skipped
		Error: UNCLASSIFIED_PROBLEM: 1: dead (exit status 134)
		(1) UNCLASSIFIED_PROBLEM
		Wraps: (2) attached stack trace
		  -- stack trace:
		  | main.glob..func14
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1173
		  | main.wrap.func1
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:281
		  | github.com/spf13/cobra.(*Command).execute
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:856
		  | github.com/spf13/cobra.(*Command).ExecuteC
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:960
		  | github.com/spf13/cobra.(*Command).Execute
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:897
		  | main.main
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:2107
		  | runtime.main
		  | 	/usr/local/go/src/runtime/proc.go:225
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1371
		Wraps: (3) 1: dead (exit status 134)
		Error types: (1) errors.Unclassified (2) *withstack.withStack (3) *errutil.leafError
```
<details><summary>Reproduce</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*tpcc/w=100/nodes=3/chaos=true.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0xdee362]

goroutine 949389 [running]:
panic(0x4ad3aa0, 0xb100620)
	/usr/local/go/src/runtime/panic.go:1065 +0x565 fp=0xc015f19b88 sp=0xc015f19ac0 pc=0x48b265
runtime.panicmem()
	/usr/local/go/src/runtime/panic.go:212 +0x5b fp=0xc015f19ba8 sp=0xc015f19b88 pc=0x4892bb
runtime.sigpanic()
	/usr/local/go/src/runtime/signal_unix.go:734 +0x173 fp=0xc015f19be0 sp=0xc015f19ba8 pc=0x4a2fb3
github.com/cockroachdb/cockroach/pkg/util/tracing.SpanFromContext(0x0, 0x0, 0x7)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/context.go:25 +0x22 fp=0xc015f19c18 sp=0xc015f19be0 pc=0xdee362
github.com/cockroachdb/cockroach/pkg/sql/execinfra.GetTraceData(0x0, 0x0, 0x4e25020, 0xc015f19c01, 0x0)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:254 +0x35 fp=0xc015f19c48 sp=0xc015f19c18 pc=0x22d9e35
github.com/cockroachdb/cockroach/pkg/sql/execinfra.GetScanStats(0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/stats.go:85 +0x4c fp=0xc015f19cb8 sp=0xc015f19c48 pc=0x22e33ac
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*joinReader).close(0xc00dde1800)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/joinreader.go:889 +0x1be fp=0xc015f19cf8 sp=0xc015f19cb8 pc=0x2732b9e
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*joinReader).ConsumerClosed(0xc00dde1800)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/joinreader.go:882 +0x2b fp=0xc015f19d10 sp=0xc015f19cf8 pc=0x27329cb
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).InternalClose(0xc00c703c00, 0x0)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:904 +0xac fp=0xc015f19d50 sp=0xc015f19d10 pc=0x22e1d8c
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Close(0xc00c703c00, 0x0, 0x0)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:289 +0x46 fp=0xc015f19d70 sp=0xc015f19d50 pc=0x2c72be6
github.com/cockroachdb/cockroach/pkg/sql/colexecop.Closers.CloseAndLogOnErr(0xc00a824480, 0x5, 0x8, 0x8b1f9d0, 0xc0115adcb0, 0xc00bc74494, 0x7)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecop/operator.go:160 +0xcc fp=0xc015f19e08 sp=0xc015f19d70 pc=0x290454c
github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc.(*Outbox).close(0xc00aabf340, 0x8b1f9d0, 0xc0115adcb0)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc/outbox.go:124 +0xc9 fp=0xc015f19e50 sp=0xc015f19e08 pc=0x2986c09
github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc.(*Outbox).Run(0xc00aabf340, 0x8b1f928, 0xc0115adcb0, 0x8a81bc0, 0xc000ae0a50, 0xb50354b100000002, 0xc0ce6bb50e47b813, 0x75ee3f7e, 0x4, 0xc00dc1c660, ...)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc/outbox.go:211 +0x2e5 fp=0xc015f19f10 sp=0xc015f19e50 pc=0x2986f45
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreator).setupRemoteOutputStream.func1(0x8b1f928, 0xc00e99cb00, 0xc00dc1c660)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:730 +0xd5 fp=0xc015f19f90 sp=0xc015f19f10 pc=0x33c88b5
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreatorHelper).accumulateAsyncComponent.func1.1(0xc00afcd950, 0x8b1f928, 0xc00e99cb00, 0xc00dc1c660, 0xc00499cd50)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1322 +0x44 fp=0xc015f19fb8 sp=0xc015f19f90 pc=0x33ca664
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1371 +0x1 fp=0xc015f19fc0 sp=0xc015f19fb8 pc=0x4c5cc1
created by github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreatorHelper).accumulateAsyncComponent.func1
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1321 +0x6f
```

--- a/pkg/sql/rowexec/joinreader.go
+++ b/pkg/sql/rowexec/joinreader.go
@@ -883,11 +883,6 @@ func (jr *joinReader) ConsumerClosed() {
 }
 
 func (jr *joinReader) close() {
-	// Make sure to clone any tracing span so that stats can pick it up later.
-	// Stats are only collected after we finish closing the processor.
-	if !jr.Closed {
-		jr.scanStats = execinfra.GetScanStats(jr.Ctx)
-	}
 	if jr.InternalClose() {
 		if jr.fetcher != nil {
 			jr.fetcher.Close(jr.Ctx)
@@ -918,6 +913,7 @@ func (jr *joinReader) execStatsForTrace() *execinfrapb.ComponentStats {
 	}
 
 	// TODO(asubiotto): Add memory and disk usage to EXPLAIN ANALYZE.
+	jr.scanStats = execinfra.GetScanStats(jr.Ctx)
 	ret := &execinfrapb.ComponentStats{
 		Inputs: []execinfrapb.InputStats{is},
 		KV: execinfrapb.KVStats{

--- a/pkg/sql/rowexec/tablereader.go
+++ b/pkg/sql/rowexec/tablereader.go
@@ -299,9 +299,6 @@ func (tr *tableReader) Next() (rowenc.EncDatumRow, *execinfrapb.ProducerMetadata
 
 func (tr *tableReader) close() {
 	if tr.InternalClose() {
-		// scanStats is collected from the trace after we finish doing work for this
-		// join.
-		tr.scanStats = execinfra.GetScanStats(tr.Ctx)
 		if tr.fetcher != nil {
 			tr.fetcher.Close(tr.Ctx)
 		}
@@ -319,6 +316,7 @@ func (tr *tableReader) execStatsForTrace() *execinfrapb.ComponentStats {
 	if !ok {
 		return nil
 	}
+	tr.scanStats = execinfra.GetScanStats(tr.Ctx)
 	ret := &execinfrapb.ComponentStats{
 		KV: execinfrapb.KVStats{
 			BytesRead:      optional.MakeUint(uint64(tr.fetcher.GetBytesRead())),

